### ISSUE TITLE
No output from page parsing, but no error messages

### ISSUE THREAD

**Issue by richard5mith** [segmentio/nightmare#224]
I've taken the cnn.js example, and my only change is changing the require line from ../nightmare to nightmare, but when I run it I only get...

> DEBUG=nightmare node --harmony cnn.js

    nightmare queueing action "goto" +0ms    
    nightmare queueing action "evaluate" +5ms

I don't ever seem to get the title of the page.

I've tried various other examples from Nightmare < 2 from around the web, and I can't get any of them to work either. Nothing ever errors, I just never seem to get any screenshot, pdf or elements pulled off the page in the output.

After 4 hours of bashing my head, I'm not sure what else to try.

**Comment by idrakimuhamad** [comment]
I tried this and got it returned properly.

    nightmare queueing action "goto" +0ms
    nightmare queueing action "evaluate" +1ms
    Breaking News, U.S., World, Weather, Entertainment & Video News - CNN.com

But if I tried with a website that have a SLL problem, like below, it hang there for, forever. 

        vo(run)(function(err, result) {
          if (err) throw err;
        });

        function *run() {
            var title = yield nightmare
            .goto('https://cas2.northport.com.my/myapp/wa/r/nmbctr')
            .evaluate(function() {
                return document.title;
            });
            console.log(title);
            yield nightmare.end();
        }

It just hang here. No error, and electron no exited.

        nightmare queueing action "goto" +0ms
        nightmare queueing action "evaluate" +1ms

**Comment by richard5mith** [comment]
What's your environment for running the cnn.js script?

Entirely headless server, or in a terminal in an existing X session, or OS X etc?

**Comment by pastawoua** [comment]
I do have the same issue on a headless server. Everything runs correctly on OSX but not on my build server.

**Comment by matthewmueller** [comment]
Yikes, major issue. 

![img](https://cldup.com/dAy_dE48-X.png)

Seems like the electron linux binding is broken or something. Anyone have any ideas?

**Comment by richard5mith** [comment]
Odd that at least you get an error message when I don't. Mine just fails silently.

My guess is that it can't get the framebuffer, which is where xvfb would normally come in. But I've tried running electron with that too and just get gtk errors.

**Comment by matthewmueller** [comment]
Okay. I'm not exactly sure how to codify this right now (or what layer to support this in), but to get it working on the official node (docker) image, you need to do the following:

```
# Install dependencies
apt-get update &&\
    apt-get install -y libgtk2.0-0 libgconf-2-4 \
    libasound2 libxtst6 libxss1 libnss3 xvfb
npm install nightmare

# Start Xvfb
Xvfb -ac -screen scrn 1280x2000x24 :9.0 &
export DISPLAY=:9.0

# Test it
apt-get install vim
vim index.js
# Paste in example
node index.js
```

More info: https://github.com/atom/electron/issues/228
Dockerfile: https://github.com/aheuermann/docker-electron

**Comment by richard5mith** [comment]
That doesn't solve it for me on my environment. Still no errors, but still no title from the CNN example either.

I had previously got Gtk3::Webkit Perl module running, which also uses xvfb, and had everything installed except libxss1 already. So I know my xvfb works.

Going round in circles trying to find suggestions. I see the same as this person when running electron through xvfb-run, https://gist.github.com/AspireToCodeBetter/130877925f52c8fb2557. 

And the Stack Overflow question points people here, http://stackoverflow.com/questions/32612868/running-electron-atom-shell-headlessly-on-linux-server-through-nightmare-js.

**Comment by matthewmueller** [comment]
Oh I don't think this is it, but can you try running master? `npm install segmentio/nightmare` ?

**Comment by richard5mith** [comment]
Hey! That did it. I can now start Xvfb on display 9 as you posted before, then I can now run the cnn example.

Looks like it was an incomplete Electron install before.

So can you modify your launch of electron process to use xvfb-run?

**Comment by matthewmueller** [comment]
@richard5mith i'm not very family with `xvfb`, by `xvfb-run` do you mean the instructions i provided? or is there a simpler way?

I think it's starting a server (`Xvfb -ac -screen scrn 1280x2000x24 :9.0 &`), which is something I'd like to avoid.

**Comment by richard5mith** [comment]
I was thinking something like this as a variation on your steps, which is how I had run Gtk Webkit .

    xvfb-run --server-args="-screen 9 1280x2000x24" ./electron

That does seem to launch correctly and the renderer process starts. So I'm not sure if you can do something similar from your JS and then do the IPC with that process.

**Comment by richard5mith** [comment]
Actually, this works, without having Xvfb running in the background first as per your previous steps...

    DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" node cnn.js

**Comment by thucnc** [comment]
Hi, I still got the issue that the nightmare hangs as the screenshot
![image](https://cloud.githubusercontent.com/assets/2088055/9975804/855ac872-5ef4-11e5-9983-cf7300e42d92.png)

I tried reinstall all the node_modules with latest nightmare package. May anyone can help ? Thansk

**Comment by pastawoua** [comment]
@richard5mith thanks for that, runs for me now on my build server!

**Comment by matthewmueller** [comment]
Nice! Now just need to figure out the best way to get these dependencies on linux boxes

**Comment by pastawoua** [comment]
@matthewmueller that might help:
https://github.com/shippableImages/minBase/blob/master/ubu1204/01base/containerSetup.sh
https://github.com/shippableImages/minBase/blob/master/ubu1204/01base/xvfb.sh

Using the shippable container to run nightmare and mocha.

**Comment by pastawoua** [comment]
@matthewmueller I've narrowed it down to `apt-get update && apt-get install -y xvfb x11-xkb-utils xfonts-100dpi xfonts-75dpi xfonts-scalable xfonts-cyrillic x11-apps clang libdbus-1-dev libgtk2.0-dev libnotify-dev libgnome-keyring-dev libgconf2-dev libasound2-dev libcap-dev libcups2-dev libxtst-dev libxss1 libnss3-dev gcc-multilib g++-multilib`

Hope that helps!

**Comment by matthewmueller** [comment]
thanks man, i think this should do it though: https://github.com/segmentio/nightmare/issues/224#issuecomment-141575361

Or are you talking about without node dependencies? 

I think the remaining items to sort out are:

- [ ] When do these installation procedures happen? Is this a `postinstall` script, a node-gyp thing, or is it something that shouldn't even be handled in this library?
- [ ] Verify that we can run ` xvfb-run --server-args="-screen 0 1024x768x24"` across multiple processes (in other words, running two of these spawn commands at once)
- [ ] Ensure that we're cleaning up after we execute (when the node process exits)

**Comment by dylanvalade** [comment]
I'm unable to fix [my #223 Heroku deploy] (https://github.com/segmentio/nightmare/issues/223). Updated to Nightmare v2.0.6 and tried custom buildpacks but couldn't sort it out.  Has anyone used xvfb on Heroku?

**Comment by matthewmueller** [comment]
Would be good if we could get some sort of post install script going so it just worked on heroku. Sorry guys, I haven't had much time to look into this lately, but hopefully we can get this sorted out soon

**Comment by notsentient** [comment]
@matthewmueller https://github.com/Rob--W/node-xvfb might be of use for controlling the xvfb server.

**Comment by dylanvalade** [comment]
@notsentient Jabin, Simply Wall St looks outstanding.  Good find on the xvfb module.

**Comment by jney** [comment]
same issue there.
@dylanvalade did you find out how to make it work ?

**Comment by dylanvalade** [comment]
@jney Unfortunately, I had to change my plan. Using Phantom now and this explains the setup.
http://stackoverflow.com/questions/22116673/cannot-find-module-phantomjs-in-heroku/32853675#32853675

After this Linux issue gets sorted out I would much rather use Nightmare.

**Comment by matthewmueller** [comment]
One thing I just realized is that nightmare is running fine on Circle CI. Probably missing something, but anyone know how that's working? I doubt they have all this additional setup, but maybe...

**Comment by edasque** [comment]
This
```
apt-get update && apt-get install -y xvfb x11-xkb-utils xfonts-100dpi xfonts-75dpi xfonts-scalable xfonts-cyrillic x11-apps clang libdbus-1-dev libgtk2.0-dev libnotify-dev libgnome-keyring-dev libgconf2-dev libasound2-dev libcap-dev libcups2-dev libxtst-dev libxss1 libnss3-dev gcc-multilib g++-multilib
```

and 
```
DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" node cnn.js
```

worked for me, might be worth updating the doc for since this is going to be mostly used on headless servers.

**Comment by danhalliday** [comment]
+1 I also have it running relatively painlessly using 'xvfb-run', after trying and failing to get it connected to a separate 'xvfb' process.

**Comment by kuba-gaj** [comment]
Did not work for me :(
Have all packages installed, using master from nightmare, headless Ubuntu 14.04.3.
No error messages, just hangs. Works fine on Mac

**Comment by dickeylth** [comment]
Not work for me either after tried install `xvfb-run`

```
➜  nightmare  DEBUG=nightmare* xvfb-run --server-args="-screen 0 1024x768x24" node --harmony test.js
  nightmare queueing action "goto" for https://www.taobao.com +0ms
  nightmare queueing action "evaluate" +6ms
```

**Comment by deinwort** [comment]
sudo yum install xorg-x11-server-Xvfb
xvfb-run --server-args="-screen 0 1024x768x24" node cnn.js

worked for me on CentOS.

**Comment by younes200** [comment]
@deinwort which version of CentOS and NodeJS are you using ?

**Comment by younes200** [comment]
On CentOS 7, with xvfb-run is not working :

```
# xvfb-run --server-args="-screen 0 1024x768x24" node --harmony cnn.js 
  nightmare queueing action "goto" for http://cnn.com +0ms
  nightmare queueing action "evaluate" +2ms
```

```
# uname -a
Linux localhost 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
````

```
# node -v 
4.2.3
```

**Comment by younes200** [comment]
Xvfb need more packages under Centos otherwise it will crash 

```
yum install xorg-x11-server-Xvfb  gtk2 libXtst GConf2 xorg-x11-fonts*
```

The best way to debug this is to run electron directly:
```
xvfb-run --server-args="-screen 0 1024x768x24 -extension RANDR" ./node_modules/nightmare/node_modules/electron-prebuilt/dist/electron
```

**Comment by matthewmueller** [comment]
Anyone get this working in a Dockerfile? Seems to be working inside the container but once I try passing in the command or entrypoint it just hangs (not at `nightmare queueing action "evaluate" +2ms`, but doesn't even run)

Here's the `Dockerfile`:

```docker
FROM node:5

# Updating ubuntu packages
RUN apt-get update

# Installing the packages needed to run Nightmare
RUN apt-get install -y \
  xvfb \
  x11-xkb-utils \
  xfonts-100dpi \
  xfonts-75dpi \
  xfonts-scalable \
  xfonts-cyrillic \
  x11-apps \
  clang \
  libdbus-1-dev \
  libgtk2.0-dev \
  libnotify-dev \
  libgnome-keyring-dev \
  libgconf2-dev \
  libasound2-dev \
  libcap-dev \
  libcups2-dev \
  libxtst-dev \
  libxss1 \
  libnss3-dev \
  gcc-multilib \
  g++-multilib

ENV DEBUG="nightmare"

# Add current directory to /app
ADD . /app

# Set current working directory as /app
WORKDIR /app

# Install npm packages
RUN npm install

# Default command. Assumes our file is index.js and our screen size is 1024x768
CMD DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" node index.js
```

**Comment by edasque** [comment]
For those not familiar with Docker, here is the step by step once you've installed it (it needs a 64 bit modern distro & kernel).

In the nightmare source directory (from this repo), create a file named Dockerfile which has the content at the bottom of this post (@matthewmueller I had to make some edits, including  replacing ```FROM node:5``` with ```FROM node:4.2``` for version pinning since 5.x is not LTS) then run:

```docker build -t nightmare-test .```

This takes a little while. Then:

```docker run nightmare-test```

expected result is:

```
Thu, 10 Dec 2015 15:03:07 GMT nightmare queueing action "goto" for http://cnn.com
Thu, 10 Dec 2015 15:03:07 GMT nightmare queueing action "evaluate"
Thu, 10 Dec 2015 15:03:07 GMT nightmare running
Breaking News, Daily News and Videos - CNN.com
Thu, 10 Dec 2015 15:03:25 GMT nightmare running
```
My Dockerfile is:
```
FROM node:4.2

# Updating ubuntu packages
RUN apt-get update

# Installing the packages needed to run Nightmare
RUN apt-get install -y \
  xvfb \
  x11-xkb-utils \
  xfonts-100dpi \
  xfonts-75dpi \
  xfonts-scalable \
  xfonts-cyrillic \
  x11-apps \
  clang \
  libdbus-1-dev \
  libgtk2.0-dev \
  libnotify-dev \
  libgnome-keyring-dev \
  libgconf2-dev \
  libasound2-dev \
  libcap-dev \
  libcups2-dev \
  libxtst-dev \
  libxss1 \
  libnss3-dev \
  gcc-multilib \
  g++-multilib

ENV DEBUG="nightmare"

# Add current directory to /app
ADD . /app

# Set current working directory as /app
WORKDIR /app

# Install npm packages
RUN npm install

# Default command. Assumes our file is index.js and our screen size is 1024x768
CMD npm install vo ; xvfb-run --server-args="-screen 0 1024x768x24" node index.js
```

My index.js for the CNN example (in the root project folder is:
```javascript
var Nightmare = require('./lib/nightmare');
var vo = require('vo');

vo(run)(function(err, result) {
  if (err) throw err;
});

function *run() {
  var nightmare = Nightmare();
  var title = yield nightmare
    .goto('http://cnn.com')
    .evaluate(function() {
      return document.title;
    });
  console.log(title);
  yield nightmare.end();
}
```

@matthewmueller, if you want me to do a PR with the docker file, an example directory documentation, let me know.

**Comment by matthewmueller** [comment]
@edasque thanks for the detailed instructions! the command i'm trying to get working is either this:

```bash
docker run nightmare xvfb-run --server-args="-screen 0 1024x768x24" node index.js
```

or 

```
docker run nightmare node index.js
```

with `xvfb-run --server-args="-screen 0 1024x768x24"` being an `ENTRYPOINT`

**Comment by edasque** [comment]
index.js being the CNN example, right?

Yes, it's a good idea to use an ENTRYPOINT for that. Let me know what I can help with.

Note however that the ENTRYPOINT will require you to install the application dependencies through nightmare's package.json in this case. Use the exec form of ENTRYPOINT so it could be overriden, maybe?

**Comment by matthewmueller** [comment]
> index.js being the CNN example, right?

Yep, precisely

> Let me know what I can help with.

If you have a chance, do you mind testing this command? `docker run nightmare xvfb-run --server-args="-screen 0 1024x768x24" node cnn.js`

I was able to get it working by keeping everything in the Dockerfile, but it'd be nice to have flexibility in what script you run. Haven't been able to get a version working where you can pass the file in.

> Note however that the ENTRYPOINT will require you to install the application dependencies through nightmare's package.json in this case. Use the exec form of ENTRYPOINT so it could be overriden, maybe?

Yah maybe best to stick with `CMD`, though you can overwrite `ENTRYPOINT` via `docker run --entrypoint`, though i'm not sure that's a great idea. I guess it depends how often you need to change the `--server-args`

**Comment by edasque** [comment]
I have the same problem, I don't know what the cause is yet. I thought it might be in the encoding of the quotes & such but I don't think so anymore

**Comment by sriducati** [comment]
After creating an app in in windows(Worked awesome) .. now its not running in fedora 64-bit.. I seriously think this is waste of time for everyone.. why dont you guyzz shut down nightmare, Its really nightmare for developers... All my time wasted, tried to fix for 2 days.... Unftunately its not giving any error messages... tried every steps....Its totally waste

**Comment by edasque** [comment]
I have had no problem here. Have you tried the Docker approach @sriducati?

**Comment by sriducati** [comment]
thats the only step left... bcos i tried almost all steps.. trust me,each and every dependencies i fixed for electron, spent 2-3 days.. but still not working...

**Comment by matthewmueller** [comment]
it works using docker. we're working on a better solution for linux. keep in mind this is free software, so please be civil. if you're unhappy with nightmare, you're free to try something else.

**Comment by sriducati** [comment]
@matthewmueller I am sorry if my comment was offencive .I was very happy with nightmare when working on windows but suddenly all my app stopped working without any error messages... its really very hard...just don't know what to do now...

**Comment by edasque** [comment]
@sriducati did you try using DEBUG=nightmare ? I think there might be a way to alter the verbosity but I can't recall.

**Comment by sriducati** [comment]
@edasque  ya i tried debug mode below is the result...

  [root@server ~]# DEBUG=nightmare node --harmony cnn.js
  nightmare queueing action "goto" for http://cnn.com +0ms
  nightmare queueing action "evaluate" +21ms
  nightmare running +3ms

**Comment by younes200** [comment]
@sriducati  `nightmare` is base on [electron](http://electron.atom.io/) which doesn't have a good support on linux speciality with headlessly window. Best way to debug this on linux is to run first electron :
`./node_modules/nightmare/node_modules/electron-prebuilt/dist/electron`

**Comment by sriducati** [comment]
@younes200 thank you, tried your method as well...result is empty

**Comment by sriducati** [comment]
Good new everyone.. nightmare worked well with ubuntu 12.04 64-bit

**Comment by younes200** [comment]
@sriducati could you please post your results here to help others too ;)

**Comment by sriducati** [comment]
@matthewmueller , @edasque , @younes200  Thank you all...After changing many operating sys version... found simple fix Ubuntu 12.04 64-bit ;) 

	root@server:~# xvfb-run node cnn.js
	Breaking News, Daily News and Videos - CNN.com
	root@server:~# node -v 
	v5.3.0
	root@server:~# npm -v
	3.3.12
	root@server:~# 

Its running in OpenVz VPS , No docker used... It would be great if anyone could replay to  https://github.com/segmentio/nightmare/issues/411 ,  I can write a complete help guide to install in VPS ...

**Comment by scarletsky** [comment]
@matthewmueller 
Why not put the solution/workaround in README ?

**Comment by naomik** [comment]
I'm also having this issue. CentOS workaround didn't work for me.

Mine doesn't hang tho, it just quits.

```
$ DEBUG=nightmare node cnn.js
  nightmare queueing action "goto" for http://cnn.com +0ms
  nightmare queueing action "evaluate" +5ms
  nightmare running +4ms
```

I tried with `xvfb` but still the same issue

```plain
$ DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" node cnn.js
  nightmare queueing action "goto" for http://cnn.com +0ms
  nightmare queueing action "evaluate" +5ms
  nightmare running +2ms
```

**Comment by adrianparr** [comment]
It would be good to get a solution to this.
I need to get Nightmare.js running on a remote headless server.

I have a Digital Ocean virtual server running Ubuntu 14.04.3 x64
I have installed Node v5.4.1

Running my app ...
`$ sudo DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" node app.js`

Gives the following output in Terminal ...
```
   nightmare queueing action "useragent" +0ms
   nightmare queueing action "viewport" +3ms
   nightmare queueing action "goto" for http://www.bbc.co.uk +4ms
   nightmare queueing action "wait" +1ms
   nightmare queueing action "screenshot" +1ms
   nightmare running +0ms
```

No screenshot is created, and the final success message (that I get when testing on localhost) is never shown.

**Comment by wascou** [comment]
Here, the same with a centOS docker image and node 5.5. :-/

**Comment by JumpeiArashi** [comment]
I worked latest nightmare on headless machine(docker node:4.2, docker-machine with virtualbox driver as host machine).
my dockerfile is
```dockerfile
FROM node:4.2

# Updating ubuntu packages
RUN apt-get update

# Installing the packages needed to run Nightmare
RUN apt-get install -y \
  xvfb \
  x11-xkb-utils \
  xfonts-100dpi \
  xfonts-75dpi \
  xfonts-scalable \
  xfonts-cyrillic \
  x11-apps \
  clang \
  libdbus-1-dev \
  libgtk2.0-dev \
  libnotify-dev \
  libgnome-keyring-dev \
  libgconf2-dev \
  libasound2-dev \
  libcap-dev \
  libcups2-dev \
  libxtst-dev \
  libxss1 \
  libnss3-dev \
  gcc-multilib \
  g++-multilib

WORKDIR /opt/test
RUN npm install vo segmentio/nightmare
```

and execute

```bash
Xvfb -ac -screen scrn 1280x2000x24 :99.0 &
export DISPLAY=99.0
node index.js
```

and index.js content is following.

```javascript
var Nightmare = require('nightmare');
var vo = require('vo');

vo(function* () {
  var nightmare = Nightmare({ show: true });
  var link = yield nightmare
    .goto('http://yahoo.com')
    .type('input[title="Search"]', 'github nightmare')
    .click('#uh-search-button')
    .wait('.ac-21th')
    .screenshot('screenshot.png')
    .evaluate(function () {
      return document.getElementsByClassName('ac-21th')[0].href;
    });
  yield nightmare.end();
  return link;
})(function (err, result) {
  if (err) return console.log(err);
  console.log(result);
});
```

but I don't want to write `"nightmare": "segmentio/nightmare"` to package.json. and my circle-ci test is failed.

**Comment by Deliaz** [comment]
Guys, I couldn't run headless nightmare based on advices from this discussion. 
So, any news? 
I tryed few docker's, and it always stuck at ``` nightmare running +1ms``` step..

**Comment by msokk** [comment]
Just a note when trying to debug - set environment variables `ELECTRON_ENABLE_LOGGING` and `ELECTRON_ENABLE_STACK_DUMPING` true.
https://github.com/atom/electron/blob/master/docs/api/environment-variables.md

Then there should be more output.

**Comment by Deliaz** [comment]
**msokk** thanks, 
but there is no new here: 
```
xvfb-run node test/index.js 
  nightmare queueing action "useragent" +0ms
  nightmare queueing action "goto" for http://google.com +5ms
  nightmare queueing action "type" +4ms
  nightmare queueing action "click" +1ms
  nightmare queueing action "wait" +1ms
  nightmare queueing action "inject" +1ms
  nightmare queueing action "evaluate" +0ms
  nightmare running +2ms
```

But it works good with the electrone: ```xvfb-run electron test/electron.js```. And I get a data, that I'm waiting for. And also see tons of logs

**Comment by huemans** [comment]
edasque's solution worked for me on Ubuntu 14.04 however my build server uses pm2 to run node. I can't get it to work with pm2. Any suggestions? I would hate to have to switch back to phantom. I've tried below:

pm2 start app.js --node-args="--expose-gc" -- DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" 

DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" pm2 start app.js  

DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" && pm2 start app.js 

pm2 start app.js --execute-command DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24" 

pm2 start DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24” app.js  

pm2 start DEBUG=nightmare xvfb-run --server-args="-screen 0 1024x768x24” app.js

**Comment by FdezRomero** [comment]
I have the same problem as @huemans, I couldn't make PM2, xvfb and Nightmare play well together.

Any ideas?

**Comment by sageeki** [comment]
Still having issues with the same issue, I tried all of your solutions, everything works perfectly on OSX but when moved to the Linux server it doesn't provide any response. It's the same environment, same versions of node and nightmare etc. On linux it will just not show any output. Any ideas?

npm@3.6.0 /usr/local/lib/node_modules/npm
Air:nm sagi$ npm -l version
{ npm: '3.6.0',
  ares: '1.10.1-DEV',
  http_parser: '2.6.2',
  icu: '56.1',
  modules: '47',
  node: '5.7.1',
  openssl: '1.0.2g',
  uv: '1.8.0',
  v8: '4.6.85.31',
  zlib: '1.2.8' }

**Comment by sageeki** [comment]
Might be a problem with electron running headless in Linux.

root@Ubuntu-1510-wily-64-minimal /tmp/electron-quick-start # xvfb-run /var/www/growthy/nm/node_modules/electron-prebuilt/dist/electron --type=gpu-process --channel=25956.0.799365973 --no-sandbox --supports-dual-gpus=false --gpu-driver-bug-workarounds=2,51,63 --disable-accelerated-video-decode --gpu-vendor-id=0x1234 --gpu-device-id=0x1111 --gpu-driver-vendor --gpu-driver-version --v8-natives-passed-by-fd --v8-snapshot-passed-by-fd
[27268:0308/205209:FATAL:scoped_file.cc(29)] Check failed: 0 == IGNORE_EINTR(close(fd)). : Bad file descriptor
#0 0x000001ae0a1e <unknown>
#1 0x000001af6bab <unknown>
#2 0x000001af716d <unknown>
#3 0x000001af08c4 <unknown>
#4 0x000001aea530 <unknown>
#5 0x000001af083e <unknown>
#6 0x000001af06a8 <unknown>
#7 0x000002b6e10d <unknown>
#8 0x000002b6e015 <unknown>
#9 0x0000010a9fa1 <unknown>
#10 0x0000010a911a <unknown>
#11 0x000002ea776c main
#12 0x7f93b3386a40 __libc_start_main
#13 0x000000539f09 <unknown>

(That's the command that node is trying to execute)

**Comment by gabrielperales** [comment]
@huemans @FdezRomero did you solve that problem? I'm trying to run my nightmare app with PM2 but I can't make it work.

**Comment by FdezRomero** [comment]
@gabrielperales No, I haven't found a solution yet. The next thing I was planning to try is the [Headless](https://www.npmjs.com/package/headless) module, a wrapper for Xvfb directly from Node. But I don't know if it will play well with Nightmare. Let me know if you try it!

**Comment by gabrielperales** [comment]
@FdezRomero Thank you for your answer. I have tried headless module... but it don't seems to work. I'll keep trying

**Comment by gabrielperales** [comment]
@FdezRomero @huemans I could do it with this module [node-xvfb](https://github.com/Rob--W/node-xvfb). I have done a test example [here](https://github.com/gabrielperales/nightmareHeadlessTest/blob/master/index.js) .

**Comment by SantoshSrinivas79** [comment]
I've got nightmare working in headless mode with code similar to the example from @gabrielperales. However, I do not think the code will work with pm2 (atleast I wasn't able to

**Comment by gabrielperales** [comment]
@SantoshSrinivas79 I have tried my code once... and I think it worked, I'll try it again later to be sure

**Comment by sageeki** [comment]
@gabrielperales - works perfectly, thank you!

**Comment by benogle** [comment]
Anyone else running this on heroku have it working? I'm getting crashing. Everything works fine locally. It works fine for websites on the smaller side, and using larger dynos occasionally lets it finish when it otherwise wouldnt indicating maybe some kind of memory thing?

```
$ node --version
v5.7.0
$ npm list | grep nightmare
+-- nightmare@2.2.0
$ DEBUG=nightmare* xvfb-run --server-args "-screen 0 1024x768x24" node cnn.js
```

I'm using all the deps outlined above. The last 3 lines of the log for running cnn.js.

```
nightmare:log did-get-response-details +3ms [{"sender":{"session":{"cookies":{},"webRequest":{}},"hostWebContents":null,"devToolsWebContents":null,"debugger":{},"_maxListeners":0,"_events":{"page-title-updated":[null,null],"crashed":[null,null],"did-finish-load":[null,null]},"_eventsCount":28,"browserWindowOptions":{"show":false,"alwaysOnTop":true,"webPreferences":{"preload":"/app/node_modules/nightmare/lib/preload.js","nodeIntegration":false}}}},false,"http://rtax.criteo.com/delivery/rta/rta.js?netId=4157&cookieName=crtg_trnr&rnd=10840509575&varName=crtg_content","http://rtax.criteo.com/delivery/rta/rta.js?netId=4157&cookieName=crtg_trnr&rnd=10840509575&varName=crtg_content",200,"GET","http://www.cnn.com/",{"cache-control":["no-cache"],"content-length":["158"],"content-type":["text/javascript; charset=utf-8"],"date":["Sat, 26 Mar 2016 01:38:45 GMT"],"expires":["Mon, 26 Jul 1997 05:00:00 GMT"],"p3p":["CP=\"NON DSP COR CURa PSA PSD OUR BUS NAV STA\""],"pragma":["no-cache"],"server":["Microsoft-IIS/8.5"],"x-powered-by":["ASP.NET"]}]
  nightmare:log did-stop-loading +118ms [{"sender":{"session":{"cookies":{},"webRequest":{}},"hostWebContents":null,"devToolsWebContents":null,"debugger":{},"_maxListeners":0,"_events":{"page-title-updated":[null,null],"crashed":[null,null],"did-finish-load":[null,null]},"_eventsCount":28,"browserWindowOptions":{"show":false,"alwaysOnTop":true,"webPreferences":{"preload":"/app/node_modules/nightmare/lib/preload.js","nodeIntegration":false}}}}]
  nightmare:log crashed +0ms [{"sender":{"session":{"cookies":{},"webRequest":{}},"hostWebContents":null,"devToolsWebContents":null,"debugger":{},"_maxListeners":0,"_events":{"page-title-updated":[null,null],"crashed":[null,null],"did-finish-load":[null,null]},"_eventsCount":28,"browserWindowOptions":{"show":false,"alwaysOnTop":true,"webPreferences":{"preload":"/app/node_modules/nightmare/lib/preload.js","nodeIntegration":false}}}}]
```

Looks like it's crashing in the renderer process. Planning on trying to upgrade electron to 0.37 and see what happens. Ideally i'd like to keep this on heroku, but maybe not in the cards. I can make a new issue if more useful.

**Comment by freeall** [comment]
@benogle Unfortunately I don't think anyone have made it work on heroku. I tried a lot, but never did make it work. Now I follow this issue to see if anyone does i the future. Maybe when Chromium gets better at running headless.

**Comment by benschwarz** [comment]
I [made a buildpack](https://github.com/benschwarz/heroku-electron-buildpack) for running this on heroku. 

:cake: Works for me. YMMV. :tada:

**Comment by gabrielperales** [comment]
Hi, I have updated [my example](https://github.com/gabrielperales/nightmareHeadlessTest) with a Dockerfile. That is all you will need to run your nightmarejs code in headless mode. I hope you can enjoy it.

**Comment by rickmed** [comment]
hey @benschwarz , how did you make it work with heroku? which dyno size were you using? mine is crashing after a certain (small) amount of action queued? I think I'm running into the whole heroku /dev/shm being too small (5 mb in heroku I believe which is useless)

**Comment by benschwarz** [comment]
@rickmed I found it to be pretty unstable too (admittedly, on the 512m dyno), I decided to use phantomjs instead. I'd love to have used electron and nightmare, but instead I shipped phantomjs and had no issues since. I'm sure there is a simple reason that this is happening on heroku environments, but I just didn't have the time to spend on it.

**Comment by taoeffect** [comment]
Think I found the "best" answer:

- [Official Electron documentation on testing headlessly](https://github.com/electron/electron/blob/master/docs/tutorial/testing-on-headless-ci.md)

It shows what to put in `.travis.yml` and recommends the super-handy [xvfb-maybe](https://github.com/paulcbetts/xvfb-maybe) utility.

Now I just need to figure out whether I should use Nightmare or [electron-mocha](https://github.com/jprichardson/electron-mocha)...

**Comment by lucasbrigida** [comment]
Simple [gist](https://gist.github.com/lucasbrigida/f9d606a6888cbd1c7874a84607486c3e) with commands for run nightmare headless.

**Comment by hhsadiq** [comment]
Is this issue still a bug in nightmare, or it was fixed in any new version..??

**Comment by rickmed** [comment]
This is not a nightmare issue but an electron (chromium really) issue. @rosshinkley maybe this should be closed? 

@hhsadiq I'll try to help you via stackoverflow

**Comment by jokeyrhyme** [comment]
FYI:

- https://chromium.googlesource.com/chromium/src/+/master/headless/README.md

- https://bugs.chromium.org/p/chromium/issues/detail?id=546953

**Comment by rosshinkley** [comment]
@rickmed I was planning on closing it with the resolution of #502 (where Nightmare would take care of the framebuffer for you, at least until Chromium's headless features trickle into Electron), but the details of that have proven much more difficult than I had anticipated.  It's not even as simple as "just use Xvfb" as [_sometimes_ this causes Chromium to hang](https://github.com/segmentio/nightmare/issues/561).

**Comment by otaviomedeiros** [comment]
# Solution with Docker

The simplest way i found to get it working with docker...

Dockerfile
```
FROM node:argon

RUN apt-get update &&\
    apt-get install -y libgtk2.0-0 libgconf-2-4 \
    libasound2 libxtst6 libxss1 libnss3 xvfb

WORKDIR /app

RUN npm install nightmare
COPY cnn.js .

COPY entrypoint /
RUN chmod +x /entrypoint
ENTRYPOINT ["/entrypoint"]

CMD DEBUG=nightmare node cnn.js
```

entrypoint

```
#!/usr/bin/env bash
set -e

# Start Xvfb
Xvfb -ac -screen scrn 1280x2000x24 :9.0 &
export DISPLAY=:9.0

exec "$@"
```

cnn.js

```
var Nightmare = require('nightmare'),
  nightmare = Nightmare();

nightmare.goto('http://cnn.com')
  .evaluate(function(){
    return document.title;
  })
  .end()
  .then(function(title){
    console.log(title);
  })
```

Build image

```
docker build -t cnn-example .
```

Then run and see the results
```
docker run cnn-example
Tue, 14 Jun 2016 13:38:35 GMT nightmare queuing process start
Tue, 14 Jun 2016 13:38:35 GMT nightmare queueing action "goto" for http://cnn.com
Tue, 14 Jun 2016 13:38:35 GMT nightmare queueing action "evaluate"
Tue, 14 Jun 2016 13:38:35 GMT nightmare running
Tue, 14 Jun 2016 13:38:51 GMT nightmare electron child process exited with code 0: success!
CNN - Breaking News, U.S., World, Weather, Entertainment & Video News
```

I'm using docker-machine with virtualbox driver

**Comment by jakeorr** [comment]
I got GAE deploys working in a simpler fashion described here: http://stackoverflow.com/a/38447642/562915

**Comment by feedm3** [comment]
I got it working on Gitlab-Ci

```
image:  node:latest

before_script:
  - apt-get update
  - apt-get install -y libgtk2.0-0 libgconf-2-4 libasound2 libxtst6 libxss1 libnss3 xvfb
  - npm install --silent
  - Xvfb -ac -screen scrn 1280x2000x24 :9.0 &
  - export DISPLAY=:9.0

build_job:
  script:
    - "npm test"
```

**Comment by mashaalmemon** [comment]
@huemans, @FdezRomero, @gabrielperales, @SantoshSrinivas79 . To get nightmare working headlessly with pm2, you just need to wrap your mind around what Xvfb does, what xvfb-run does, how they interact with your node process, and how this interaction is facilitated.

Going to give you an idea of the basics and then tell you how to get this all going in pm2. Forgive me if I am oversimplifying things:
- Xvfb is a frame buffer which allows us to mock a desktop interface where, in this case, the electron browser can run.
- Xvfb sets a display number (the initial default is "99"). The application using the desktop must have an idea of this display number to use the frame buffer.
- xvfb-run (which is a shell script) is perfect for single process node applications as what it actually does (as far as I understand) is setup the Xvfb process, set the "DISPLAY" environment variable to ":99", and then spawn the process you specify with access to the environment variable so it can access the display.
- For PM2 to manage processes happily it must spawn all the processes that it controls. I think this is where xvfb-run falls apart with pm2. It creates processes that need to be controlled and then exits and pm2 doesn't like this.
- Instead what we can do is spawn two processes simultaneously (do what xvfb-run does essentially): one being Xvfb and the other being a process for your application, setting the display number and "DISPLAY" environment variable explicitly.
- The best way to spawn the two processes is to create a PM2 Process file json as explained [here](http://pm2.keymetrics.io/docs/usage/application-declaration/) in pm2 docs, and then use the pm2 start command.

This is what your process file would look like, substituting for your specific case:

```json
{
  "apps" : [{
    "name"        : "<your_app_name>",
    "script"      : "<your_main.js>",
    "env": {
      "DISPLAY": ":99"
    }
  },
    {
      "name"        : "Xvfb",
      "interpreter" : "none",
      "script"      : "Xvfb",
      "args"        : ":99"
    }]
}
```

If you place this process file in the same directory as `<your_main.js>`, and call the file `process.json` then you would execute this pm2 command to get the two processes started:

```
pm2 start process.json
```

Tested and working on Debian 8.

**Comment by atorralb** [comment]
Everybody is going astray, nobody is giving a solution. This is the oneliner that will save your day ( obviously if you have xfvb installed... modify to your needs.. this one is for slackware 14.1). 

*Note, on some systems is xvfb is called "Xvfb" and in other "xvfb-run"

**DEBUG=nightmare*  Xvfb :99 -screen 0 1024x768x24 > /dev/null 2>&1 & node app.js**

**Comment by stevenvachon** [comment]
I'm not much of a terminal user, and while I tried Electron's travis-ci recommendations, my tests are still timing out, even with 10+ second limits.

**Comment by maxfriedmann** [comment]
I got an image working based on this discussion, but when I create a screenshot, I don't see any text. Selecting any text via nightmare works though. Anyone ever seen that issue?

**Comment by myeongsic89** [comment]
Centos 7 isn't work?

**Comment by pelhage** [comment]
From a performance POV, does anyone know how Nightmare running with Xvfb compares to PhantomJS as headless?

**Comment by stevenvachon** [comment]
@pelhage well, Electron sure doesn't take a few *seconds* to open like PhantomJS does.

**Comment by atorralb** [comment]
@pelhage I haven't used phantomJS since Nightmare decided to bundle electron instead of webkit.  I think there are many reason you should choose Electron instead of phantom's webkit such as but not limited to : 
-Browser technology up to date with standards (unlike phantomJS webkit)
-Electron is becoming a defacto standard for automation testing and it will probably take over Selenium

**Comment by pelhage** [comment]
I used @mashaalmemon 's solution for my Scraper cron job, and it worked for the first day, but ever since I've been getting error messages in my log and my scraper won't run. Not sure how to decipher them, can someone maybe point me in the right direction or tell me what typically causes these errors?

Gist of logs https://gist.github.com/pelhage/3f43240a3bd36a7ec06cb09d06c3be92

**Comment by entrptaher** [comment]
The ultimate solution.
`pm2 start --interpreter xvfb-run npm -- start`

**Comment by stevenmiller888** [comment]
Has anyone successfully gotten Nightmare to work on Heroku? If so, can you share your Dockerfile?

cc @rosshinkley

**Comment by yoz** [comment]
@stevenmiller888 Take a look at https://github.com/captain401/heroku-buildpack-xvfb. You'll also want heroku-buildpack-apt as in the instructions.

(You might also need to include libnotify4 and libxss1 in your Aptfile - let me know if I need to update the instructions.)

**Comment by Dachow** [comment]
My website was deployed on EC2, and I tried `sudo DEBUG=nightmare xvfb-run --server-args "-screen 0 1024x768x24" node app.js`, but it processes the request only once
![image](https://cloud.githubusercontent.com/assets/11835233/20952971/36eee936-bc6b-11e6-8cd0-7c59196ec66b.png)

**Comment by GautierT** [comment]
@otaviomedeiros : Hi ! Did u manage to launch multiple long script at the same time with this config ?
I tried and when i launch 4-5 times the same script on the same xvfb display it's hang and i got some nightmare .wait() timeout. 
Same if i launch multiple xvfb display and attach one script to one display. 
Seems that a docker instance can't handle many xvfb display... 
No error output from nightmare. Just a `Error: .wait() timed out after 180000msec` at last

**Comment by GautierT** [comment]
@Dachow : it's seems that you killed the electron child process before the other request had time to finish. show us your code !

**Comment by Dachow** [comment]
@GautierT  Hi! I have ran it with express. The basic structure of code like this:
```
router.get('/', function (req, res, next) {
    nightmare
        .useragent('...')
        .goto('https://www.instagram.com/')
        .click('._aj7mu._taytv._ki5uo._o0442')
        .insert('form input[type="email"]', 'xxx@gmail.com')
        .insert('form input[type="password"]', 'xxx')
        .click('form button')
        .wait('article')
        .evaluate(function () {
            // return something
        }
        .end()
        .then(function(result){
            // some handlers
            res.render('body', {
                items: items
            });
        })
        .catch(function (error) {
            console.error('Search failed:', error);
        });
}
```
It processes the request only once

**Comment by entrptaher** [comment]
@Dachow, that's because you are ending the instance on first run.

**Comment by jekku** [comment]
@Dachow I think you can make a new instance everytime the route is called, if your resources can handle it so

**Comment by mashaalmemon** [comment]
Folks following my original post on August 11, 2016 I have some more information to share. I noticed some others had similar problem that I just identified. 

Going back to my post you'll see I mentioned how to get nightmare.js working on linux via the use of pm2 and Xvfb. This still applies to a linux bare metal of vm installation.

However to run nightmare in a docker a few things to keep in mind:
-  You'll need to run your node.js script and Xvfb in the docker simultaneously. [pm2-docker](http://pm2.keymetrics.io/docs/usage/docker-pm2-nodejs/) is a good option for running multiple processes in a docker.
- I noticed that some folks had this all working in a docker except screenshots. Or they had situations where things were working in some cases but were not. Dockers have by default 64MB shared memory (/dev/shm). Electron (chromium) needs more shared memory so you'll need to increase this. Look into docker documentation to find out more but it can be done either as part of your "docker run" command or in your .yml files. See issue #587 .
- Lastly electron processes that are complete stick around as [Zombie processes](https://www.google.ca/amp/s/blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/amp/?client=safari). There are solutions although I haven't decided on which one I'll use, but it is a problem that should be addressed.

Hopefully one or more of these points helps you folks.

**Comment by mashaalmemon** [comment]
@GautierT, no need to have multiple Xvfb displays in my experience. One Xvfb display and as many simultaneous nightmare processes as your system can handle reasonably.

**Comment by adek05** [comment]
One more thing that I encountered when trying to run using xvfb-run. Apparently, if a process executing Xvfb has pid 1, then Xvfb does not send SIGUSR1 which is what xvfb-run relies on.
More info here: https://codereview.chromium.org/1398053003/
Workaround could be as simple as running `/bin/bash -c "xvfb-run <command>`

**Comment by sandstrom** [comment]
Since the ~2 weeks ago Chromium support headless mode.

> With the latest Chrome Canary on Linux you can enable headless mode with the --headless command line flag. (https://bugs.chromium.org/p/chromium/issues/detail?id=546953#c148)

@matthewmueller @rosshinkley  Will this change in Chromium make any difference here?

Nightmare is awesome, and a headless mode (without xvfb) would be very useful! 🥇 

*(here is a related issue in Electron https://github.com/electron/electron/issues/228)*

**Comment by dickeylth** [comment]
@sandstrom That's really cool! Looking forward to see it working in nightmare!

**Comment by Chris911** [comment]
Heroku just bumped the size of `/dev/shm/` on their dyno to allow headless browser test 🎉.

https://devcenter.heroku.com/changelog-items/1085

**Comment by abcfy2** [comment]
I used `xvfb-run`, but got these error:

```
$ DEBUG=nightmare*,electron* xvfb-run --server-args="-screen 0 1280x2000x24" npm test -- --progress=false --single-run --watch=false

> linchang@0.0.0 test /var/lib/jenkins/workspace/customer-client-wx
> ng test --code-coverage "--progress=false" "--single-run" "--watch=false"


START:
(node:12768) DeprecationWarning: loaderUtils.parseQuery() received a non-string value which can be problematic, see https://github.com/webpack/loader-utils/issues/56
parseQuery() will be replaced with getOptions() in the next major version of loader-utils.
03 03 2017 15:24:45.811:WARN [karma]: Port 9876 in use
03 03 2017 15:24:45.813:INFO [karma]: Karma v1.5.0 server started at http://0.0.0.0:9877/
03 03 2017 15:24:45.814:INFO [launcher]: Launching browser Nightmare with unlimited concurrency
03 03 2017 15:24:45.819:INFO [launcher]: Starting browser Nightmare
Fri, 03 Mar 2017 07:24:45 GMT nightmare queuing process start
Fri, 03 Mar 2017 07:24:45 GMT nightmare queueing action "goto" for http://localhost:9877/?id=7995405
Fri, 03 Mar 2017 07:24:45 GMT nightmare queueing action "wait"
Fri, 03 Mar 2017 07:24:45 GMT nightmare queueing action "evaluate"
Fri, 03 Mar 2017 07:24:45 GMT nightmare queueing action "wait"
Fri, 03 Mar 2017 07:24:45 GMT nightmare running

Fri, 03 Mar 2017 07:24:46 GMT electron:stderr Xlib:  extension "RANDR" missing on display ":99".

Fri, 03 Mar 2017 07:24:46 GMT electron:stderr Xlib:  extension "RANDR" missing on display ":99".

Fri, 03 Mar 2017 07:24:46 GMT electron:stderr [12818:0303/152446.458888:FATAL:platform_font_linux.cc(63)] Check failed: typeface. Could not find any font: Nimbus Sans L, sans
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #0 0x000001ccd51e <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #1 0x000001cb27cb <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #2 0x000003b48e3d <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #3 0x000003b48a7f <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #4 0x000003b49736 <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #5 0x000003b562f9 <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #6 0x000002a668ce <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #7 0x00000293abd3 <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #8 0x000003c13654 atom::api::WebContents::InitWithSessionAndOptions()
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #9 0x000003c13dad atom::api::WebContents::WebContents()
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #10 0x000003c1dd16 atom::api::WebContents::Create()
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #11 0x000003c2f24f atom::api::Window::Window()
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #12 0x000003c30b6c atom::api::Window::New()
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #13 0x000003bece00 mate::internal::InvokeNew<>()
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #14 0x000003becf90 mate::internal::Dispatcher<>::DispatchToCallback()
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #15 0x7fc91b24ceff <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #16 0x7fc91adf744b <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #17 0x7fc91adf6f09 <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr #18 0x2764fc7043a7 <unknown>
Fri, 03 Mar 2017 07:24:46 GMT electron:stderr 

Fri, 03 Mar 2017 07:24:47 GMT nightmare electron child process exited with code null: undefined

Fri, 03 Mar 2017 07:24:47 GMT nightmare electron child process not started yet, skipping kill.

03 03 2017 15:24:47.152:ERROR [launcher]: Cannot start Nightmare
	
03 03 2017 15:24:47.156:INFO [launcher]: Trying to start Nightmare again (1/2).
Fri, 03 Mar 2017 07:24:48 GMT nightmare queuing process start
```

**Comment by plesiecki** [comment]
@abcfy2 Man read the log there is a clue. I'll give you a hint `FATAL:platform_font_linux.cc(63)] Check failed: typeface. Could not find any font: Nimbus Sans L, sans`. Peace ;)

**Comment by abcfy2** [comment]
It's a little strange. I find the `Nimbus Sans L` font exists. But after install some other fonts(I don't remember which font solved my issue, maybe `apt install -y fonts-noto`?), it works.

**Comment by matthewmueller** [comment]
Hey folks, I'm going to lock this conversation since there's a ton of good information available and each comment emails 54 people.

If you have an issue with getting Nightmare working on Electron please first read through the comments here. If you cannot find an answer, open a new issue. Thanks!

**Comment by matthewmueller** [comment]
I want to leave an example here in case anyone comes to this thread later and finds it hard to follow. Here's a minimal way to get it running on ubuntu (using Docker):

**Dockerfile**

```dockerfile
FROM node:8.3.0

RUN apt-get update

# Installing the packages needed to run Nightmare
RUN apt-get install -y \
  xvfb \
  x11-xkb-utils \
  xfonts-100dpi \
  xfonts-75dpi \
  xfonts-scalable \
  xfonts-cyrillic \
  x11-apps \
  clang \
  libdbus-1-dev \
  libgtk2.0-dev \
  libnotify-dev \
  libgnome-keyring-dev \
  libgconf2-dev \
  libasound2-dev \
  libcap-dev \
  libcups2-dev \
  libxtst-dev \
  libxss1 \
  libnss3-dev \
  gcc-multilib \
  g++-multilib

RUN npm install nightmare xvfb
```

**index.js**

```js
const Nightmare = require('nightmare')
const Xvfb = require('xvfb')

main().catch(console.error)

// main function
async function main() {
  const close = await xvfb()
  const nightmare = Nightmare()

  const [err, title] = await poss(run(nightmare))
  if (err) {
    // cleanup properly
    await nightmare.end()
    await close()
    throw err
  }

  console.log(title)

  // shut'er down
  await nightmare.end()
  await close()
}

// run nightmare
//
// put all your nightmare commands in here
async function run(nightmare) {
  await nightmare.goto('https://google.com')
  const title = await nightmare.title()
  return title
}

// xvvb wrapper
function xvfb(options) {
  var xvfb = new Xvfb(options)

  function close() {
    return new Promise((resolve, reject) => {
      xvfb.stop(err => (err ? reject(err) : resolve()))
    })
  }

  return new Promise((resolve, reject) => {
    xvfb.start(err => (err ? reject(err) : resolve(close)))
  })
}

// try/catch helper
async function poss(promise) {
  try {
    const result = await promise
    return [null, result]
  } catch (err) {
    return [err, null]
  }
}
```

**Makefile**

```Makefile
build:
	@docker build -t nightmare .

run:
	@docker run -it -v $(PWD)/index.js:/index.js nightmare node index.js

run.debug:
	@docker run -it -v $(PWD)/index.js:/index.js -e DEBUG=nightmare* nightmare node index.js
```

**Usage**

```sh
make build    # build the docker container
make run      # run the container (`make run.debug` to see debug logs)
```

Hope this helps! 🎉
